File: app/src/main/java/ru/magflayer/spectrum/presentation/pages/main/toolbar/ToolbarViewHolder.java
package ru.magflayer.spectrum.presentation.pages.main.toolbar;

import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.arellomobile.mvp.MvpDelegate;
import com.arellomobile.mvp.presenter.InjectPresenter;

import ru.magflayer.spectrum.presentation.pages.main.router.MainRouter;
import ru.magflayer.spectrum.presentation.pages.main.router.MainRouterImpl;

public class ToolbarViewHolder implements ToolbarView {

    @InjectPresenter
    ToolbarPresenter presenter;

    private Toolbar toolbar;
    private MvpDelegate<ToolbarViewHolder> mvpDelegate;

    public ToolbarViewHolder(final AppCompatActivity activity, final Toolbar toolbar) {
        this.toolbar = toolbar;
        getMvpDelegate().onCreate();
        getMvpDelegate().onAttach();

        MainRouter mainRouter = new MainRouterImpl(activity);
        presenter.setRouter(mainRouter);

        activity.setSupportActionBar(toolbar);
        ActionBar actionBar = activity.getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            toolbar.setNavigationOnClickListener(view -> presenter.handleBack());
        }
    }

    @Override
    public void showToolbar() {
        toolbar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideToolbar() {
        toolbar.setVisibility(View.GONE);
    }

    @Override
    public void showTitle(final String title) {
        toolbar.setTitle(title);
    }

    public void onDestroy() {
        getMvpDelegate().onDetach();
        getMvpDelegate().onDestroy();
    }

    private MvpDelegate getMvpDelegate() {
        if (mvpDelegate == null) {
            mvpDelegate = new MvpDelegate<>(this);
        }
        return mvpDelegate;
    }
}

File: app/src/main/java/ru/magflayer/spectrum/presentation/pages/main/toolbar/ToolbarViewHolder.kt
package ru.magflayer.spectrum.presentation.pages.main.toolbar

import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.android.components.ActivityComponent
import moxy.MvpDelegate
import moxy.presenter.InjectPresenter
import moxy.presenter.ProvidePresenter

class ToolbarViewHolder(
    private val activity: AppCompatActivity,
    private val toolbar: Toolbar
) : ToolbarView {

    @InjectPresenter
    lateinit var presenter: ToolbarPresenter

    private var mvpDelegate = MvpDelegate(this)

    init {
        getMvpDelegate().onCreate()
        getMvpDelegate().onAttach()

        activity.setSupportActionBar(toolbar)
    }

    @EntryPoint
    @InstallIn(ActivityComponent::class)
    interface ToolbarEntryPoint {
        fun presenter(): ToolbarPresenter
    }

    @ProvidePresenter
    fun providePresenter(): ToolbarPresenter {
        return EntryPointAccessors.fromActivity(
            activity,
            ToolbarEntryPoint::class.java
        ).presenter()
    }

    override fun showToolbar() {
        toolbar.visibility = View.VISIBLE
    }

    override fun hideToolbar() {
        toolbar.visibility = View.GONE
    }

    override fun showTitle(title: String) {
        toolbar.title = title
    }

    fun onDestroy() {
        getMvpDelegate().onDetach()
        getMvpDelegate().onDestroy()
    }

    private fun getMvpDelegate(): MvpDelegate<ToolbarViewHolder> {
        return mvpDelegate
    }
}
